### PR TITLE
feat(compiler): warning if unsupported package manager is used

### PR DESCRIPTION
**Pull request by Chriscbr**
Adding a small warning to help users encountering issues using winglibs with other JavaScript package managers. (https://github.com/winglang/wing/issues/6129). Screenshot of what the diagnostic looks like:

Screenshot 2024-07-29 at 7 34 50 PM

I ended up also removing some code for injecting panics into the compiler that has seldom been used in my experience (it introduced some snapshot regressions and removing the code seemed like the simpler fix).

## Checklist

- [ ] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [ ] Description explains motivation and solution
- [ ] Tests added (always)
- [ ] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: libs/wingc/src/lib.rs
Comment (MarkMcCulloh) [inline libs/wingc/src/lib.rs:161]: nit: `const`

File: libs/wingc/src/lib.rs
Comment (MarkMcCulloh) [inline libs/wingc/src/lib.rs:166]: nit: `const`, or maybe `static` since it's a `String`

File: libs/wingc/src/lib.rs
Comment (MarkMcCulloh) [inline libs/wingc/src/lib.rs:294]: To avoid the cost of multiple syscalls for the average case, what if this check only happens if there are errors? Ideally, we'd only want to check this if we specifically fail package resolution.

File: libs/wingc/src/parser.rs
Comment (MarkMcCulloh) [inline libs/wingc/src/parser.rs:2261]: 🥳

File: libs/wingc/src/lib.rs
Comment (MarkMcCulloh) [inline libs/wingc/src/lib.rs:160]: Inverting the check for the success case (look for a `package-lock.json` instead) could be useful since there's only one case that is supposed to work and many cases that might not.

A `package-lock.json` file may not be present in cases where there are no dependencies, but ideally we shouldn't even be checking this in that case.

**Review by MarkMcCulloh — APPROVED** [review]
Left some comments but I understand we probably don't want to overcomplicate this so it LGTM as-is 👍

File: libs/tree-sitter-wing/grammar.js
Comment (eladb) [inline libs/tree-sitter-wing/grammar.js:466]: If I recall the whole idea of this "feature" (or backdoor) was to test that panics are handled properly and diagnostics are reported for them.

Maybe reconsider removing them.

**Review by eladb — APPROVED** [review]

File: libs/wingc/src/lib.rs
Comment (Chriscbr) [reply libs/wingc/src/lib.rs:160]: > could be useful since there's only one case that is supposed to work and many cases that might not

I guess I'm OK with the possibility that this list might become outdated (since it's easy to add onto). Like you said, if the user has no dependencies (or maybe they just aren't saving lock files), package-lock.json might not be present so I'm not sure how we'd avoid the check otherwise.

Comment (MarkMcCulloh) [reply libs/wingc/src/lib.rs:160]: It's not only about outdated cases, but cases this doesn't handle like workspaces since the lockfile would only present at the root, not necessarily next to a package.json

File: libs/wingc/src/lib.rs
Comment (Chriscbr) [reply libs/wingc/src/lib.rs:294]: > Ideally, we'd only want to check this if we specifically fail package resolution.

Perhaps. But even if you use pnpm initially and only add a winglib later, getting the warning earlier in your journey as a first-time user might be better, I'd argue.

For now I'm optimizing for the pain of users being unable to compile their Wing apps, since that's a critical point where users can drop off (they might get confused by the error, have no idea how to fix it, and give up before asking on Discord or GitHub). Minimizing the number of warnings for happy Wing users who insist on using a different package manager is a secondary issue.

File: libs/tree-sitter-wing/grammar.js
Comment (MarkMcCulloh) [reply libs/tree-sitter-wing/grammar.js:466]: Could be interesting to instead elevate this past a debug feature and turn it into an intrinsic ala `panic!` or `todo!`

File: libs/wingc/src/lib.rs
Comment (Chriscbr) [reply libs/wingc/src/lib.rs:164]: Mmm I see. Yeah, fair enough, we can probably do more here (maybe do a lookup for the project's node_modules and check if there's a .pnpm folder, or only raise the error when a module lookup error occurs like you suggested) -- definitely open to improving it.

Let's merge this in for now, if you wanna try and fine tune the warning then I'll totes take the contribution

File: libs/tree-sitter-wing/grammar.js
Comment (Chriscbr) [reply libs/tree-sitter-wing/grammar.js:466]: Noted.

(I think there are better ways to test the custom panic handler / CLI error message that don't require inserting special syntax and plumbing it through all of the stages of the compiler)

**Pull request merged**
